### ISSUE TITLE
[Enhancement] Remove container history on deletion

### ISSUE THREAD

**Issue by bb010g** [stoically/firefox-add-on-temporary-containers#17]
I'd like to use this as a full replacement for [Private Tab](https://addons.mozilla.org/en-US/firefox/addon/private-tab/), but the history sticking around makes clean-up annoying. It would be nice if you could set history to be deleted for a temporary container.

**Comment by stoically** [comment]
Hi, unfortunately this is currently [not supported](https://github.com/mozilla/multi-account-containers/issues/47) by Firefox, history entries don't have a [tabid or containerid](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/history/HistoryItem). If it gets supported one day I'll happily implement it too.

**Comment by anewuser** [comment]
I've opened this to see if they give any feedback: https://github.com/mozilla/multi-account-containers/issues/1096 .

If they don't reply soon, at least consider deleting history entries created for websites entered in the "Always per Website" option of your extension, in the same fashion as https://addons.mozilla.org/firefox/addon/historia/ . It constantly monitors the history to delete all URLs you specify, but to make it use less resources, you could make it cleanup the history only after the last temporary container is closed.

**Comment by stoically** [comment]
I could add a way to automatically delete history like Historia does - by default it only listens to the [history.onVisited event](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/history/onVisited) and removes the HistoryItem immediately if one of the added domain patterns matches. That's already light on resources. The patterns could indeed be the ones added to "Always per Website".

Or I could a preference to automatically remove all urls from history that got opened in a temporary container.

But in both cases it would affect the *complete* history, not only the history of temporary containers - so you could end up deleting history that you actually wanted to preserve - so I'm not exactly sure if I want to add such a feature until it's officially supported, especially since there's already the Historia Add-on that can do the job just fine.

**Comment by stoically** [comment]
After thinking about it a bit - here's what I plan to implement.

![image](https://user-images.githubusercontent.com/29637501/35769185-6bd32088-0907-11e8-9d52-f141701d8f50.png)

**Comment by stoically** [comment]
Just published version 0.29 with Delete History feature. Hope you like it. Please let me know if something doesn't work as expected.

**Comment by anewuser** [comment]
The cleanup doesn't work if you disable "Automatic Mode" and use the <kbd>Alt</kbd> + <kbd>P</kbd> shortcut.

Can you add an option to use no-history containers with the extension button and context menu too?

**Comment by stoically** [comment]
Ooops. Thanks for letting me know. Just published version 0.30 with the fix. Lets just pretend 0.29 never happened. 😬 

Yeah, the toolbar icon and context menu should open deletes-history containers too when the "Automatically create" configuration under advanced is set. I'll fix that in the next version.

**Comment by anewuser** [comment]
The fix worked, but Firefox asked me to give it permission to access the history again after the update. Maybe there's a bug that causes that when one first allowed access to something using the method you used in the Advanced pane?

**Comment by stoically** [comment]
Yeah, happened to me too. Seems to be an issue with Firefox; updates after that don't ask for the permission anymore. But I guess it's not too bad because it will only happen to people that explicitly wanted to give that permission anyway.

Another thing I just noticed - clearing history for a site isn't enough to prevent that it pops up as suggestion in the address bar (have to select it and click `del`) and the [omnibox API](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/omnibox) doesn't have a way to delete something. That's kind of a bummer.

**Comment by anewuser** [comment]
I was going to report it on Bugzilla, but then saw that you've deleted 0.29. 

The problem is that after an automatic update users may dismiss the popup without reading it (and then have the extension disabled or stuck in an old version), or people may just start accepting all requests from all extensions without thinking twice (https://palant.de/2016/07/02/why-mozilla-shouldn-t-copy-chrome-s-permission-prompt-for-extensions), so the fewer times the permissions popup appear, the better. 

Hopefully your users will be more careful, but fixing this problem would be good for Firefox extensions in general.

**Comment by stoically** [comment]
Yeah, I've deleted the version as soon as I saw your report and started fixing after that. But just published 0.31 with the fix for the "Automatically create" thing. So if you'd want, you could report now. That'd be great, thanks!

I guess you're absolutely right and the article makes sense. Also I didn't know that Add-ons stop updating if you reject a permission prompt, that's bad indeed.

**Comment by stoically** [comment]
In case you report the bug, could you post the link here? Or let me know if you ain't got the time to do so. After your explanation I think it actually is important to report.

**Comment by anewuser** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1435617

**Comment by stoically** [comment]
Awesome, thanks.

**Comment by stoically** [comment]
Just published 0.32 with a note about that bug including the link in the advanced preferences.

**Comment by SegFault0x0** [comment]
@stoically This bugged me for a while until I looked at your code.  It appears that it's doing exactly what it's supposed/written to do.  In `src/background/container.js` line 305 (`tryToRemove` method), you have:
```javascript
if (tempTabs.length > 0) {
        debug('[tryToRemove] not removing container because it still has tabs', cookieStoreId, tempTabs.length);
        return;
}
```
So as long as I leave *one* tab open for a container, I can close and re-open as many other tabs under that same container, whether they be in the same [sub]domain or opened with ALT+X.  It's not until I close the last one that all of the recent tab history is deleted.  I think this is good, as it works similar to Private Browsing for a specific container (i.e. Closing a container's last tab is like closing a PB window).

However, it'd be nice to have an option (either in Preferences, or perhaps, a popup) that asks whether or not we wish to delete a container when closing its last tab.  If "no", then the container could be deleted whenever the browser window closes.
